# Урок 1. Оформление ноутбука и закрепление функций и генераторов

## Задача 1.  
***На складе лежат разные фрукты в разном количестве. Нужно написать функцию, которая на вход принимает любое количество названий фруктов и их количество, а возвращает общее количество фруктов на складе***

**Решение**

*Задаём список кортежей. Первое поле - название фрукта, второе - количество на складе*

In [9]:
fruits = [('Apple',15),('Banana',12),('Pear',17),('Plum',28),('Orange',32)]
print(fruits)

[('Apple', 15), ('Banana', 12), ('Pear', 17), ('Plum', 28), ('Orange', 32)]


*Теперь описываем функцию, принимающую на вход список кортежей, и возвращающую их общее количество на складе*

In [10]:
def count_fruits(fruits_list:list):
    count = 0
    for i in range(len(fruits_list)):
        count += fruits_list[i][1]
    return count

*И находим сумму из списка*

In [11]:
print(count_fruits(fruits))

104


*Проверяем: 15 + 12 + 17 + 28 + 32 = 104. Функция считает верно*

## Задача 2.  
***Дан список с затратами на рекламу. Но в данных есть ошибки, некоторые затраты имеют отрицательную величину. Удалите такие значения из списка и посчитайте суммарные затраты.  
[100, 125, -90, 345, 655, -1, 0, 200]  
Используйте list comprehensions***

**Решение**

*Задаём исходный список затрат:*

In [18]:
expenses = [100, 125, -90, 345, 655, -1, 0, 200]

*Удалаяем отрицательные элементы, используя list comprehensions:*

In [19]:
expenses = [item for item in expenses if item >= 0]

*Проверяем полученный список. Он не должен содержать отрицательные элементы:*

In [17]:
print(expenses)

[100, 125, 345, 655, 0, 200]


*И находим сумму затрат*

In [20]:
print(sum(expenses))

1425


*Ответ: 1425*

*Можно реализовать более сжато:*

In [21]:
expenses = [100, 125, -90, 345, 655, -1, 0, 200]
print(sum([item for item in expenses if item >= 0]))

1425


*Или даже так:*

In [22]:
print(sum([item for item in [100, 125, -90, 345, 655, -1, 0, 200] if item >= 0]))

1425


*Как видим, ответ во всех случаях одинаков*

## Задача 3

***Даны два списка.  
Дата покупки  
['2021-09-14', '2021-12-15', '2021-09-08', '2021-12-05', '2021-10-09', '2021-09-30', '2021-12-22', '2021-11-29', '2021-12-24', '2021-11-26', '2021-10-27', '2021-12-18', '2021-11-09', '2021-11-23', '2021-09-27', '2021-10-02', '2021-12-27', '2021-09-20', '2021-12-13', '2021-11-01', '2021-11-09', '2021-12-06', '2021-12-08', '2021-10-09', '2021-10-31', '2021-09-30', '2021-11-09', '2021-12-13', '2021-10-26', '2021-12-09']  
Суммы покупок по датам  
[1270, 8413, 9028, 3703, 5739, 4095, 295, 4944, 5723, 3701, 4471, 651, 7037, 4274, 6275, 4988, 6930, 2971, 6592, 2004, 2822, 519, 3406, 2732, 5015, 2008, 316, 6333, 5700, 2887]***

***1. Найдите, какая выручка у компании в ноябре. Используйте list comprehensions***

**Решение**

*Задаём два списка:*

In [40]:
dates = ['2021-09-14', '2021-12-15', '2021-09-08', '2021-12-05', '2021-10-09', '2021-09-30', '2021-12-22', '2021-11-29', '2021-12-24', '2021-11-26', '2021-10-27', '2021-12-18', '2021-11-09', '2021-11-23', '2021-09-27', '2021-10-02', '2021-12-27', '2021-09-20', '2021-12-13', '2021-11-01', '2021-11-09', '2021-12-06', '2021-12-08', '2021-10-09', '2021-10-31', '2021-09-30', '2021-11-09', '2021-12-13', '2021-10-26', '2021-12-09']
sums = [1270, 8413, 9028, 3703, 5739, 4095, 295, 4944, 5723, 3701, 4471, 651, 7037, 4274, 6275, 4988, 6930, 2971, 6592, 2004, 2822, 519, 3406, 2732, 5015, 2008, 316, 6333, 5700, 2887]

*Проверяем длины списков, они должны совпадать*

In [29]:
print('Даты  : ',len(dates))
print('Суммы : ',len(sums))

Даты  :  30
Суммы :  30


*Здесь необходимо преобразовать список **dates** из строкового типа в тип дата/время*

*Для начала подключим соответствующий модуль:*

In [31]:
from datetime import datetime

*Теперь преобразуем строки в числа:*

In [41]:
dates = [datetime.strptime(data, "%Y-%m-%d").date() for data in dates]

*Проверяем что получилось:*

In [42]:
print(dates)

[datetime.date(2021, 9, 14), datetime.date(2021, 12, 15), datetime.date(2021, 9, 8), datetime.date(2021, 12, 5), datetime.date(2021, 10, 9), datetime.date(2021, 9, 30), datetime.date(2021, 12, 22), datetime.date(2021, 11, 29), datetime.date(2021, 12, 24), datetime.date(2021, 11, 26), datetime.date(2021, 10, 27), datetime.date(2021, 12, 18), datetime.date(2021, 11, 9), datetime.date(2021, 11, 23), datetime.date(2021, 9, 27), datetime.date(2021, 10, 2), datetime.date(2021, 12, 27), datetime.date(2021, 9, 20), datetime.date(2021, 12, 13), datetime.date(2021, 11, 1), datetime.date(2021, 11, 9), datetime.date(2021, 12, 6), datetime.date(2021, 12, 8), datetime.date(2021, 10, 9), datetime.date(2021, 10, 31), datetime.date(2021, 9, 30), datetime.date(2021, 11, 9), datetime.date(2021, 12, 13), datetime.date(2021, 10, 26), datetime.date(2021, 12, 9)]


*Теперь обрабатываем по заданному условию - находим выручку компании в ноябре:*

In [43]:
summa = 0
for item in range(len(dates)):
    if dates[item].month == 11 :
        summa += sums[item]
print(summa)

25098


*Ответ: 25098*

***2. Найдите выручку компании в зависимости от месяца
Для этого напишите функцию, которая на вход принимает список с датами и список с выручкой, а на выходе словарь, где ключи - это месяцы, а значения - это выручка.
Используйте аннотирование типов.***

*Создаём функцию:*

In [46]:
def get_sums(dat:list,sm:list)->list:
    ans : dict = {}
    for item in range(len(dat)):
        if dat[item].month in ans:
            ans[dat[item].month] = ans[dat[item].month] + sm[item]
        else:
            ans.update({dat[item].month : sm[item]})
    return ans

*Получаем результат:*

In [53]:
dct = get_sums(dates,sums)

*И смотрим что получилось:*

In [54]:
for key in dct:
    print(key, dct[key])

9 25647
12 45452
10 28645
11 25098


*Всё здорово, но... Месяцы идут в разнобой, по мере их добавления в словарь. Поэтому словарь следует отсортировать по ключу:*

In [55]:
dct = dict(sorted(dct.items()))

In [59]:
for key in dct:
    print(key, dct[key])

9 25647
10 28645
11 25098
12 45452


*Здесь мы видим, что выручка за ноябрь та же, что и ответ в предыдущем задании*